### PR TITLE
Add proof output IDs

### PR DESCRIPTION
**Pull request by chris124567**
https://github.com/SiaFoundation/explored/issues/124

A little annoyed that I had to redefine FileContractElement in `explorer`.  Alternatively could have gone with a custom marshaler in `api` but it would've recovered similar amounts of new types but with more aliasing so I decided this was the way to go.

### CODE REVIEW

File: explorer/types.go [hunk 1]
```diff
@@ -123,9 +123,38 @@ type SiafundOutput struct {
...
+	Payout             types.Currency          `json:"payout"`
+	ValidProofOutputs  []ContractSiacoinOutput `json:"validProofOutputs"`
+	MissedProofOutputs []ContractSiacoinOutput `json:"missedProofOutputs"`
+	UnlockHash         types.Hash256           `json:"unlockHash"`
+	RevisionNumber     uint64                  `json:"revisionNumber"`
+}
+
+// A FileContractElement is a record of a FileContract within the state
+// accumulator.
+type FileContractElement struct {
```
Comment (n8maninger) [inline explorer/types.go:149]: I think you can drop `FileContractElement` from the FileContract type. You should be able to replace it with just the ID and get everything necessary.

```
// A FileContract is a storage agreement between a renter and a host. It
// contains a bidirectional payment channel that resolves as either "valid" or
// "missed" depending on whether a valid StorageProof is submitted for the
// contract.
type FileContract struct {
	ID                 types.FileContractID    `json:"id"`
	Filesize           uint64                  `json:"filesize"`
	FileMerkleRoot     types.Hash256           `json:"fileMerkleRoot"`
	WindowStart        uint64                  `json:"windowStart"`
	WindowEnd          uint64                  `json:"windowEnd"`
	Payout             types.Currency          `json:"payout"`
	ValidProofOutputs  []ContractSiacoinOutput `json:"validProofOutputs"`
	MissedProofOutputs []ContractSiacoinOutput `json:"missedProofOutputs"`
	UnlockHash         types.Hash256           `json:"unlockHash"`
	RevisionNumber     uint64                  `json:"revisionNumber"`
}

// A FileContractRevision is a FileContract with extra fields for revision
// information.
type FileContractRevision struct {
	ParentID           types.FileContractID    `json:"parentID"`
	UnlockConditions   types.UnlockConditions  `json:"unlockConditions"`
	Filesize           uint64                  `json:"filesize"`
	FileMerkleRoot     types.Hash256           `json:"fileMerkleRoot"`
	WindowStart        uint64                  `json:"windowStart"`
	WindowEnd          uint64                  `json:"windowEnd"`
	ValidProofOutputs  []ContractSiacoinOutput `json:"validProofOutputs"`
	MissedProofOutputs []ContractSiacoinOutput `json:"missedProofOutputs"`
	UnlockHash         types.Hash256           `json:"unlockHash"`
	RevisionNumber     uint64                  `json:"revisionNumber"`
}
```

Comment (chris124567) [reply explorer/types.go:149]: How do you feel about the `StateElement`?  The `LeafIndex` is useful for retrieving the `MerkleProof`.

Comment (n8maninger) [reply explorer/types.go:149]: leaf index and merkle proof are unnecessary for v1 file contracts

File: explorer/types.go [hunk 2]
```diff
@@ -123,9 +123,32 @@ type SiafundOutput struct {
...
-// A FileContract is a types.FileContractElement with added fields for
-// resolved/valid state.
+// A ContractSiacoinOutput is a SiacoinOutput with an added field for its ID.
+type ContractSiacoinOutput struct {
+	ID types.SiacoinOutputID `json:"id"`
+	types.SiacoinOutput
+}
+
+// A FileContract is a storage agreement between a renter and a host. It
+// contains a bidirectional payment channel that resolves as either "valid" or
```
Comment (ChrisSchinnerl) [inline explorer/types.go:133]: I'd rephrase this a bit. Instead of `contains` say `serves as` since the contract is the channel and strictly speaking its only safe to use as a unidirectional payment channel since otherwise the host has incentive to not publish the latest revision which should be the one that contains the highest host payout.

File: explorer/types.go [hunk 3]
```diff
@@ -123,9 +123,32 @@ type SiafundOutput struct {
...
+	Payout             types.Currency          `json:"payout"`
+	ValidProofOutputs  []ContractSiacoinOutput `json:"validProofOutputs"`
+	MissedProofOutputs []ContractSiacoinOutput `json:"missedProofOutputs"`
+	UnlockHash         types.Hash256           `json:"unlockHash"`
+	RevisionNumber     uint64                  `json:"revisionNumber"`
+}
+
+// A EnhancedFileContract is a FileContractElement with added fields for
+// resolved/valid state, and when the transaction was confirmed and proved.
+type EnhancedFileContract struct {
```
Comment (ChrisSchinnerl) [inline explorer/types.go:151]: Maybe `ExtendedFileContract` is a better name? Otherwise it feels like it should be better ^^

File: explorer/types.go (hunk 2 above)
Comment (chris124567) [reply explorer/types.go]: I changed it here but also want to say I copied the docstring from `core/types`.

**Review by ChrisSchinnerl — APPROVED** [review]
